sdk/go/src/sawtooth_sdk/client/encoder.go

```diff
@@ -0,0 +1,156 @@
+package client
+
+import (
+	"encoding/hex"
+	"fmt"
+	"github.com/golang/protobuf/proto"
+	"sawtooth_sdk/protobuf/batch_pb2"
+	"sawtooth_sdk/protobuf/transaction_pb2"
+	"strings"
+	"time"
+)
+
+// See sdk/go/tests/client_test.go for example usage.
+
+type Transaction struct{ pb *transaction_pb2.Transaction }
+type Batch struct{ pb *batch_pb2.Batch }
+
+type TransactionEncoder struct {
+	privkey         []byte
+	pubkey          []byte
+	familyName      string
+	familyVersion   string
+	payloadEncoding string
+	addNonce        bool
+}
+
+// NewTransactionEncoder constructs a new encoder which can be used to generate
+// transactions and batches, and to serialize batches for submitting to the
+// REST API.
+func NewTransactionEncoder(privkey []byte, familyName, familyVersion, payloadEncoding string, addNonce bool) *TransactionEncoder {
+	return &TransactionEncoder{
+		privkey:         privkey[:],
+		pubkey:          GenPubKey(privkey),
+		familyName:      familyName,
+		familyVersion:   familyVersion,
+		payloadEncoding: payloadEncoding,
+		addNonce:        addNonce,
+	}
+}
+
+// NewTransaction Creates a new transaction and handles the construction and
+// signing of the transaction header.
+func (self *TransactionEncoder) NewTransaction(payload []byte, inputs, outputs, dependencies []string) Transaction {
+	hash := Encode(SHA512(payload), "hex")
+
+	header := &transaction_pb2.TransactionHeader{
+		BatcherPubkey:   Encode(self.pubkey, "hex"),
+		FamilyName:      self.familyName,
+		FamilyVersion:   self.familyVersion,
+		Inputs:          inputs,
+		Outputs:         outputs,
+		Dependencies:    dependencies,
+		PayloadEncoding: self.payloadEncoding,
+		PayloadSha512:   hash,
+		SignerPubkey:    Encode(self.pubkey, "hex"),
+	}
+
+	if self.addNonce {
+		header.Nonce = fmt.Sprintf("%x", time.Now().UTC().UnixNano())
+	}
+
+	headerBytes, err := proto.Marshal(header)
+	if err != nil {
+		panic(err)
+	}
+	headerSig := Encode(Sign(headerBytes, self.privkey), "hex")
+
+	return Transaction{
+		pb: &transaction_pb2.Transaction{
+			Header:          headerBytes,
+			HeaderSignature: headerSig,
+			Payload:         payload,
+		},
+	}
+}
+
+// GetId Returns the Transaction ID which can be used to specify this
+// transaction as a dependency for other transactions.
+func (self Transaction) Id() string {
+	return self.pb.GetHeaderSignature()
+}
+
+// NewBatch creates a new batch from the given transactions created by
+// NewTransaction. It handles the construction and signing of the batch header.
+func (self *TransactionEncoder) NewBatch(transactions []Transaction) Batch {
```

Comment:
I agree on the first point.

I disagree on the second point. Rather than overloading the interface (which in Go means either type casting or defining a new method) I think it makes more sense to provide a method for parsing the serialized TransactionList, which can then be passed to the NewBatch() method.